packaging/scripts/postremove.sh

```diff
@@ -35,8 +35,9 @@ if [ "$1" = "upgrade" ]; then
 	fi
 
 	if [ $INTERACTIVE -eq 1 ]; then
-	    checkConfigOutput=$(/usr/bin/evcc checkconfig 2>&1 || true)
-		if ! echo "$checkConfigOutput" | grep -q "config valid"; then
+    	checkConfigOutput=$(bash -c '/usr/bin/evcc checkconfig 2>&1; echo Returncode: $?')
+
+		if ! echo "$checkConfigOutput" | grep -q "Returncode: 0"; then
```

Comment:
Ich habe tatsächlich wenig Ahnung von Shellskripting, aber wäre das nicht viel einfacher so:

    if ! $(/usr/bin/evcc checkconfig 2>&1); then

Reply:
> Ich habe tatsächlich wenig Ahnung von Shellskripting, aber wäre das nicht viel einfacher so:
> 
> ```
> if ! $(/usr/bin/evcc checkconfig 2>&1); then
> ```

Das sieht so zwar schöner aus, funktioniert aber leider nicht richtig. Das Problem ist, dass dpkg den Fehlercode von checkconfig erkennt und bereits das Skript als fehlgeschlagen ansieht und abbricht, bevor der if oder else body ausgeführt wird. Mit ||true oder der Kapselung in einer eigenen Bash lässt sich der Fehlschlag verbergen.  Da der Returncode der Bash aber immer 0 sein sollte, ist es schwierig, den Returncode direkt auszulesen. Deshalb das Echo/Grep.

Reply:
> Das Problem ist, dass dpkg den Fehlercode von checkconfig erkennt und bereits das Skript als fehlgeschlagen ansieht und abbricht, bevor der if oder else body ausgeführt wird.

Ich kann das nicht kommentieren, kommt mir aber wirklich komisch vor- das ist ja auch nur ein Shellskript. Wie macht dpkg das???

Reply:
Das sagt Github Copilot falls das hilft:

GitHub Copilot
The dpkg command is used in Debian-based Linux distributions to handle packages. If the /usr/bin/evcc checkconfig command internally uses dpkg or any other command that produces output to stderr on success, the handling would differ between the two code variants.

In the first variant:

`checkConfigOutput=$(bash -c '/usr/bin/evcc checkconfig 2>&1; echo Returncode: $?')
if ! echo "$checkConfigOutput" | grep -q "Returncode: 0"; then`

The output and error messages from the dpkg command (or any other command) are captured in the checkConfigOutput variable. The exit status of the command is also captured. If the command fails, or if it succeeds but produces output to stderr, the grep command will not find "Returncode: 0" in checkConfigOutput, and the code inside the if block will be executed.

In the second variant:

`if ! $(/usr/bin/evcc checkconfig 2>&1); then`

Only the exit status of the command is checked. If the command fails, the code inside the if block will be executed. However, if the command succeeds but produces output to stderr, this will not cause the if block to be executed, because the exit status of the command is still 0 (indicating success). The output and error messages from the command are not captured or checked.

Reply:
> Wie macht dpkg das???

Dafür sorgt das `set -e` am Anfang des Skripts.

Reply:
Um genau zu sein erkennt das also auch nicht dpkg, sondern die Shell die dieses Skript ausführt, und die dann sofort aussteigt sobald ein Prozess RC!=0 hat.

Reply:
Warst schneller. Habe es gerade herausgefunden :)
Default ist set +e => (Fehler beenden Skript nicht)
Eingestellt ist aber set-e (Gute Praxis für Installationsskripte)

Was ich vorher falsch verstanden hatte:
Die Fehlermeldung wird von Dpkg gekapselt, es hat aber eigentlich mit einem Standardfeature der Bash zu tun. 
Ich pushe gleich eine leicht angepasste Version.

Wieder was gelernt 👍

Reply:
Mein Laienverständnis war immer, dass `set -e` *unbehandelte* Fehler abfängt. Darum geht es hier aber nicht. Kleiner Test:

```sh
#!/bin/sh
set -e

if ! $(/bin/sh -c false); then
  echo "This should be printed"
fi

```
funktioniert wie erwartet mit Ausgabe.

Reply:
@nicok1 ...und man kann sich nicht darauf verlassen, dass es `bash` überhaupt gibt ;)

Reply:
Stimmt. Ich denke dann ist die von @andig vorgeschlagene Vereinfachung möglich.

Reply:
Du hast recht bezüglich der Bash. Ich habe den Code nochmal angepasst. Ich möchte den Output des Befehls im Fehlerfall mit zurückgeben, ohne den Befehl bestenfalls zwei mal auszuführen zu müssen.

Reply:
Ok. Neuer Versuch. Jetzt wird keine status Variable verwendet. Jetzt muss dafür aber der Output des Befehls unterdrückt werden. set +e ist nicht erforderlich in diesem Fall.